### PR TITLE
add watch_dir command

### PR DESCRIPTION
**Pull request by evalexpr**
Adds a command named `watch_dir` that watches the directory passed to it
recursively

### CODE REVIEW

**Comment by evalexpr** [comment]
The only way I could find to test this was to use `~/.config/direnv/direnvrc`

**Comment by zimbatm** [comment]
:+1: for the idea, but the implementation should be recursive.

**Comment by zimbatm** [comment]
I think something like this would do it, but it requires to also add `shopt -s globstar` to the script:  https://github.com/koalaman/shellcheck/wiki/Sc2044#if-find-is-just-matching-globs-recursively

**Comment by evalexpr** [comment]
Hmm, it seems that the macos tests are using bash < 4.0 which doesn't include the shopt globstar. I think macos ships with zsh by default now but it'd be nice to make it work with that older version of bash too - maybe we need to switch to a `find` approach.

I've also tested this on a directory with 7000 files in, and it seemed to hang - are we concerned about users watching a 7000 file directory?

**Comment by zimbatm** [comment]
Good points. The downside with the `find` approach is that it adds an external dependency, which is why I proposed globstar.

Alternatively, `watch_dir` could be implemented in Go directly. Go has a [path/filepath.Walk](https://golang.org/pkg/path/filepath/#Walk) function in the stdlib that should make the implementation relatively straight-forward. And it should be more efficient than spawning 7000 watch_file.

**Comment by evalexpr** [comment]
Definitely faster

```
❯ time direnv reload
direnv reload  0.00s user 0.00s system 110% cpu 0.003 total

❯ ./direnv status | wc -l
7241
```

Any tips on a more thorough test would be appreciated :)

File: cmd_watch_dir.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+		err = watches.Unmarshal(watchString)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
```
Comment (zimbatm) [inline cmd_watch_dir.go:55]:
```suggestion
if info.Isdir() {
  return nil
}
```
We should only be watching files as folders don't have meaningful mtimes.

File: cmd_watch_dir.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	"fmt"
+	"os"
+	"path/filepath"
+)
+
+// CmdWatchDir is `direnv watch-dir SHELL PATH`
+var CmdWatchDir = &Cmd{
+	Name:    "watch-dir",
+	Desc:    "Recursively adds a directory to the list that direnv watches for changes",
+	Args:    []string{"[SHELL]", "[DIR]"},
```
Comment (zimbatm) [inline cmd_watch_dir.go:13]:
```suggestion
	Args:    []string{"SHELL", "DIR"},
```
Let's make both arguments required to simplify the logic. Potentially we want to add a variable number of directories in the future to watch multiple folders at once.

**Comment by zimbatm** [comment]
Almost ready!

If you want to unpack the gzenv format, direnv comes with a tool builtin: `direnv show_dump "$DIRENV_WATCHES"`

File: cmd_watch_dir.go (hunk 1 above)
Comment (evalexpr) [reply cmd_watch_dir.go:55]: Is it not worth keeping directories as part of this? meaning that we'll reload `direnv` whenever we add/remove a file from the directory (or a subdirectory). I'm not sure how mtimes work so I'm probably wrong here :)

Comment (evalexpr) [reply cmd_watch_dir.go:48]: I have added this suggestion for now though

File: test/direnv-test-common.sh [hunk 3]
```diff
@@ -201,6 +203,21 @@ test_start "failure"
...
+test_start "watch-dir"
+    echo "No watches by default"
+    test_eq "${DIRENV_WATCHES}" "${WATCHES}"
+
+    direnv_eval
+
+    if ! direnv show_dump "${DIRENV_WATCHES}" | grep -q "testfile"; then
+        echo "FAILED: testfile not added to DIRENV_WATCHES"
+        exit 1
+    fi
```
Comment (evalexpr) [inline test/direnv-test-common.sh:215]: I wasn't sure about adding this here, is it worth it?

File: cmd_watch_dir.go (hunk 1 above)
Comment (zimbatm) [reply cmd_watch_dir.go:48]: Actually, I don't know.

I assumed that the directory didn't have a meaningful mtime. But now that I am testing it locally, it looks like it *does* update whenever a file gets created. So I might be wrong. It looks like it works on Linux, but I don't know if it's portable across POSIX.

File: test/direnv-test-common.sh (hunk 3 above)
Comment (zimbatm) [reply test/direnv-test-common.sh:215]: Yeah, sure! Thanks for adding tests.

File: test/direnv-test-common.sh [hunk 4]
```diff
@@ -92,6 +92,8 @@ test_start base
   direnv_eval
   echo "${HELLO}"
   test -z "${HELLO}"
+
+  unset WATCHES
```
Comment (zimbatm) [inline test/direnv-test-common.sh:96]:
```suggestion
  unset DIRENV_WATCHES
```
Is that what you meant?

File: cmd_watch_dir.go (hunk 1 above)
Comment (yermulnik) [reply cmd_watch_dir.go:48]: Not only it does but it also has to: https://serverfault.com/a/388100 (and yep looks like POSIX is what IEEE 1003 is).

File: test/direnv-test-common.sh (hunk 4 above)
Comment (evalexpr) [reply test/direnv-test-common.sh]: Nope, we set `WATCHES` on line 77:

```
WATCHES=$DIRENV_WATCHES
```

Meaning that we need to unset it at the end of this test I think. Sorry I've not had much time to look at this PR

**Comment by zimbatm** [comment]
@evalexpr do you mind checking why the checks are failing?

**Comment by evalexpr** [comment]
@zimbatm I've left a comment on the previous review comment: https://github.com/direnv/direnv/pull/697#discussion_r537421210 as to what I _think_ the reason is.

I will have time to look into it further this weekend if necessary - after that we can decide whether we want to support watching the directory itself too. Do we still think that is a bad idea?

Having the directory as part of the watches seemed to work for me locally (again I can try testing on macOS this weekend - maybe we can merge without that and create a follow-up PR)

**Comment by evalexpr** [comment]
@zimbatm I have opted to remove the `isDir` check due to the fact that it looks like we _can_ rely on mtimes for directories. I have tested this on both NixOS and macOS by creating a file and directory inside both the top level of a directory tracked via `watch_dir` as well as deep in a subdirectory, all of these scenarios cause direnv to reload which is a desirable outcome I think.

This should be good to go now :)

**Comment by zimbatm** [comment]
thanks for doing this, I already reached the same conclusion and was going to remove IsDir as well :)

**Pull request merged**